File: bridges/punchd-bridge/gateway/src/auth/dpop.ts
/**
 * DPoP Proof Verification (RFC 9449) for the gateway.
 *
 * Checks: typ, alg, signature, htm, htu, iat freshness, jti replay, cnf.jkt binding.
 */

import { createPublicKey, createHash, verify } from "crypto";

function base64UrlEncode(buf: Buffer): string {
  return buf.toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
}

function base64UrlDecode(str: string): Buffer {
  let s = str.replace(/-/g, "+").replace(/_/g, "/");
  while (s.length % 4) s += "=";
  return Buffer.from(s, "base64");
}

/** Compute JWK Thumbprint (RFC 7638) — SHA-256 of the canonical JWK */
function computeJwkThumbprint(jwk: any): string {
  let canonical: string;
  if (jwk.kty === "EC") {
    canonical = `{"crv":"${jwk.crv}","kty":"${jwk.kty}","x":"${jwk.x}","y":"${jwk.y}"}`;
  } else if (jwk.kty === "OKP") {
    canonical = `{"crv":"${jwk.crv}","kty":"${jwk.kty}","x":"${jwk.x}"}`;
  } else if (jwk.kty === "RSA") {
    canonical = `{"e":"${jwk.e}","kty":"${jwk.kty}","n":"${jwk.n}"}`;
  } else {
    throw new Error(`Unsupported key type: ${jwk.kty}`);
  }
  const hash = createHash("sha256").update(canonical).digest();
  return base64UrlEncode(hash);
}

// JTI replay protection — track seen JTIs with TTL
const seenJtis = new Map<string, number>();
const JTI_TTL_MS = 120_000; // 2 minutes

function checkAndStoreJti(jti: string): boolean {
  const now = Date.now();
  // Purge expired entries periodically
  if (seenJtis.size > 1000) {
    seenJtis.forEach((exp, k) => {
      if (exp < now) seenJtis.delete(k);
    });
  }
  if (seenJtis.has(jti)) return false;
  seenJtis.set(jti, now + JTI_TTL_MS);
  return true;
}

export interface DPoPVerifyResult {
  valid: boolean;
  error?: string;
}

/**
 * Verify a DPoP proof JWT (RFC 9449).
 */
export function verifyDPoPProof(
  proofJwt: string,
  httpMethod: string,
  httpUrl: string,
  expectedJkt?: string,
): DPoPVerifyResult {
  try {
    const parts = proofJwt.split(".");
    if (parts.length !== 3) return { valid: false, error: "Invalid JWT structure" };

    const header = JSON.parse(base64UrlDecode(parts[0]).toString());
    const payload = JSON.parse(base64UrlDecode(parts[1]).toString());

    // Check typ
    if (header.typ !== "dpop+jwt") return { valid: false, error: "Invalid typ" };

    // Check alg — support EdDSA and ES256/384/512
    const supportedAlgs = ["EdDSA", "ES256", "ES384", "ES512"];
    if (!supportedAlgs.includes(header.alg)) return { valid: false, error: `Unsupported alg: ${header.alg}` };

    // Must have jwk in header
    if (!header.jwk) return { valid: false, error: "Missing jwk in header" };

    // Import the public key and verify signature
    const publicKey = createPublicKey({ key: header.jwk, format: "jwk" });
    const signInput = `${parts[0]}.${parts[1]}`;
    const signature = base64UrlDecode(parts[2]);

    const alg = header.alg === "EdDSA" ? null : header.alg.toLowerCase().replace("es", "sha");
    const valid = verify(alg, Buffer.from(signInput), publicKey, signature);
    if (!valid) return { valid: false, error: "Invalid signature" };

    // Check htm (HTTP method)
    if (payload.htm !== httpMethod) return { valid: false, error: `htm mismatch: ${payload.htm} != ${httpMethod}` };

    // Check htu (HTTP URL, without query string)
    const expectedHtu = httpUrl.split("?")[0];
    if (payload.htu !== expectedHtu) return { valid: false, error: "htu mismatch" };

    // Check iat freshness (allow 2 minute skew)
    const now = Math.floor(Date.now() / 1000);
    if (Math.abs(now - payload.iat) > 120) return { valid: false, error: "iat too far from current time" };

    // Check jti replay
    if (!payload.jti || !checkAndStoreJti(payload.jti)) return { valid: false, error: "jti missing or replayed" };

    // Check cnf.jkt binding if provided
    if (expectedJkt) {
      const thumbprint = computeJwkThumbprint(header.jwk);
      if (thumbprint !== expectedJkt) return { valid: false, error: "JWK thumbprint does not match cnf.jkt" };
    }

    return { valid: true };
  } catch (err) {
    return { valid: false, error: `DPoP verification error: ${err}` };
  }
}

/**
 * Extract cnf.jkt from a JWT payload (without full verification — the JWT
 * should already be verified by the JWKS verifier).
 */
export function extractCnfJkt(token: string): string | undefined {
  try {
    const parts = token.split(".");
    if (parts.length !== 3) return undefined;
    const payload = JSON.parse(base64UrlDecode(parts[1]).toString());
    return payload?.cnf?.jkt;
  } catch {
    return undefined;
  }
}

File: bridges/punchd-bridge-rs/src/auth/dpop.rs
use std::time::{SystemTime, UNIX_EPOCH};

use dashmap::DashMap;
use sha2::{Digest, Sha256};

use super::tidecloak::{b64url_decode, b64url_encode};

static JTI_TTL_MS: u64 = 120_000;

pub struct DPoPVerifier {
    seen_jtis: DashMap<String, u64>,
}

impl DPoPVerifier {
    pub fn new() -> Self {
        Self {
            seen_jtis: DashMap::new(),
        }
    }

    fn check_and_store_jti(&self, jti: &str) -> bool {
        let now_ms = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        if self.seen_jtis.len() > 1000 {
            self.seen_jtis.retain(|_, exp| *exp > now_ms);
        }
        if self.seen_jtis.contains_key(jti) {
            return false;
        }
        self.seen_jtis.insert(jti.to_string(), now_ms + JTI_TTL_MS);
        true
    }

    pub fn verify_proof(
        &self,
        proof_jwt: &str,
        http_method: &str,
        http_url: &str,
        expected_jkt: Option<&str>,
    ) -> Result<(), String> {
        let parts: Vec<&str> = proof_jwt.split('.').collect();
        if parts.len() != 3 {
            return Err("Invalid JWT structure".into());
        }

        let header: serde_json::Value =
            serde_json::from_slice(&b64url_decode(parts[0])?)
                .map_err(|e| format!("Header: {e}"))?;
        let payload: serde_json::Value =
            serde_json::from_slice(&b64url_decode(parts[1])?)
                .map_err(|e| format!("Payload: {e}"))?;
        let sig = b64url_decode(parts[2])?;

        if header["typ"].as_str() != Some("dpop+jwt") {
            return Err("Invalid typ".into());
        }

        let alg = header["alg"].as_str().ok_or("Missing alg")?;
        if !["EdDSA", "ES256", "ES384", "ES512"].contains(&alg) {
            return Err(format!("Unsupported alg: {alg}"));
        }

        let jwk = header.get("jwk").ok_or("Missing jwk in header")?;

        // Verify signature
        let sign_input = format!("{}.{}", parts[0], parts[1]);
        let kty = jwk["kty"].as_str().ok_or("Missing kty")?;
        let valid = match (kty, alg) {
            ("OKP", "EdDSA") => {
                let x = jwk["x"].as_str().ok_or("Missing x")?;
                super::tidecloak::b64url_decode(x)
                    .and_then(|x_bytes| {
                        let pk = ring::signature::UnparsedPublicKey::new(
                            &ring::signature::ED25519,
                            &x_bytes,
                        );
                        Ok(pk.verify(sign_input.as_bytes(), &sig).is_ok())
                    })?
            }
            ("EC", alg) => {
                let x = jwk["x"].as_str().ok_or("Missing x")?;
                let y = jwk["y"].as_str().ok_or("Missing y")?;
                let alg_ring = match alg {
                    "ES256" => &ring::signature::ECDSA_P256_SHA256_FIXED,
                    "ES384" => &ring::signature::ECDSA_P384_SHA384_FIXED,
                    _ => return Err(format!("Unsupported EC alg: {alg}")),
                };
                let x_bytes = b64url_decode(x)?;
                let y_bytes = b64url_decode(y)?;
                let mut point = Vec::with_capacity(1 + x_bytes.len() + y_bytes.len());
                point.push(0x04);
                point.extend_from_slice(&x_bytes);
                point.extend_from_slice(&y_bytes);
                let pk = ring::signature::UnparsedPublicKey::new(alg_ring, &point);
                pk.verify(sign_input.as_bytes(), &sig).is_ok()
            }
            _ => return Err(format!("Unsupported DPoP key/alg: {kty}/{alg}")),
        };

        if !valid {
            return Err("Invalid signature".into());
        }

        // Check htm
        if payload["htm"].as_str() != Some(http_method) {
            return Err("htm mismatch".into());
        }

        // Check htu
        let expected_htu = http_url.split('?').next().unwrap_or(http_url);
        if payload["htu"].as_str() != Some(expected_htu) {
            return Err("htu mismatch".into());
        }

        // Check iat
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();
        let iat = payload["iat"].as_u64().ok_or("Missing iat")?;
        if now.abs_diff(iat) > 120 {
            return Err("iat too far from current time".into());
        }

        // Check jti
        let jti = payload["jti"].as_str().ok_or("jti missing")?;
        if !self.check_and_store_jti(jti) {
            return Err("jti replayed".into());
        }

        // Check thumbprint
        if let Some(expected) = expected_jkt {
            let thumbprint = compute_jwk_thumbprint(jwk)?;
            if thumbprint != expected {
                return Err("JWK thumbprint does not match cnf.jkt".into());
            }
        }

        Ok(())
    }
}

pub fn compute_jwk_thumbprint(jwk: &serde_json::Value) -> Result<String, String> {
    let kty = jwk["kty"].as_str().ok_or("Missing kty")?;
    let canonical = match kty {
        "EC" => format!(
            r#"{{"crv":"{}","kty":"{}","x":"{}","y":"{}"}}"#,
            jwk["crv"].as_str().ok_or("Missing crv")?,
            kty,
            jwk["x"].as_str().ok_or("Missing x")?,
            jwk["y"].as_str().ok_or("Missing y")?,
        ),
        "OKP" => format!(
            r#"{{"crv":"{}","kty":"{}","x":"{}"}}"#,
            jwk["crv"].as_str().ok_or("Missing crv")?,
            kty,
            jwk["x"].as_str().ok_or("Missing x")?,
        ),
        "RSA" => format!(
            r#"{{"e":"{}","kty":"{}","n":"{}"}}"#,
            jwk["e"].as_str().ok_or("Missing e")?,
            kty,
            jwk["n"].as_str().ok_or("Missing n")?,
        ),
        other => return Err(format!("Unsupported key type: {other}")),
    };
    Ok(b64url_encode(&Sha256::digest(canonical.as_bytes())))
}

pub fn extract_cnf_jkt(token: &str) -> Option<String> {
    let parts: Vec<&str> = token.split('.').collect();
    if parts.len() != 3 {
        return None;
    }
    let payload_bytes = b64url_decode(parts[1]).ok()?;
    let payload: serde_json::Value = serde_json::from_slice(&payload_bytes).ok()?;
    payload["cnf"]["jkt"].as_str().map(|s| s.to_string())
}
